firedrake/function.py

```diff
@@ -487,3 +500,197 @@ def __idiv__(self, expr):
             assemble_expressions.IDiv(self, expr))
 
         return self
+
+    @utils.cached_property
+    def ctypes(self):
+        # Retrieve data from Python object
+        function_space = self.function_space()
+        mesh = function_space.mesh()
+        coordinates = mesh.coordinates
+        coordinates_space = coordinates.function_space()
+
+        # Store data into ``C struct''
+        c_function = _CFunction()
+        c_function.n_cols = mesh.num_cells()
+        if hasattr(mesh, '_layers'):
+            c_function.n_layers = mesh.layers - 1
+        else:
+            c_function.n_layers = 1
+        c_function.coords = coordinates.dat.data.ctypes.data_as(POINTER(c_double))
+        c_function.coords_map = coordinates_space.cell_node_list.ctypes.data_as(POINTER(c_int))
+        c_function.f = self.dat.data.ctypes.data_as(POINTER(c_double))
+        c_function.f_map = function_space.cell_node_list.ctypes.data_as(POINTER(c_int))
+        c_function.sidx = mesh.spatial_index and mesh.spatial_index.ctypes
+
+        # Return pointer
+        return ctypes.pointer(c_function)
+
+    @utils.cached_property
+    def _c_evaluate(self):
+        result = make_c_evaluate(self)
+        result.restype = int
+        return result
+
+    def evaluate(self, coord, mapping, component, index_values):
+        # Called by UFL when evaluating expressions at coordinates
+        return self.at(coord)
+
+    def at(self, arg, *args, **kwargs):
+        """Evaluate function at points."""
+        from mpi4py import MPI
+        comm = MPI.COMM_WORLD
+
+        if args:
+            arg = (arg,) + args
+        arg = np.array(arg, dtype=float)
+
+        dont_raise = kwargs.get('dont_raise', False)
+
+        # Handle f.at(0.3)
+        if not arg.shape:
+            arg = arg.reshape(-1)
+
+        # Immersed not supported
+        tdim = self.function_space().mesh().ufl_cell().topological_dimension()
+        gdim = self.function_space().mesh().ufl_cell().geometric_dimension()
+        if tdim < gdim:
+            raise NotImplementedError("Point is almost certainly not on the manifold.")
+
+        # Validate geometric dimension
+        if arg.shape[-1] == gdim:
+            pass
+        elif len(arg.shape) == 1 and gdim == 1:
+            arg = arg.reshape(-1, 1)
+        else:
+            raise ValueError("Point dimension (%d) does not match geometric dimension (%d)." % (arg.shape[-1], gdim))
+
+        # Check if we have got the same points on each process
+        root_arg = comm.bcast(arg, root=0)
+        same_arg = arg.shape == root_arg.shape and np.allclose(arg, root_arg)
+        diff_arg = comm.allreduce(int(not same_arg), op=MPI.SUM)
```

Comment:
I note that if we cared about efficiency, it might be worthwhile switching to the buffer (numpy array) version of these: e.g (`comm.Bcast` over `comm.bcast`).  But you probably don't care for now.